### PR TITLE
storage: improve the under-replicated range metric

### PR DESCRIPTION
**Pull request by petermattis**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/11983)

### CODE REVIEW

**Comment by petermattis** [comment]
screen shot 2016-12-03 at 1 41 38 pm

The spike on the far left is what our under-replicated metric looked like before this PR for a 4m node outage. The sharp drop when the node revived was not realistic. The other spikes show 2m, 4m and 6m outages with this PR. I don't have an explanation for the rapid fall after the 6m PR.

**Comment by tamird** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/storage/store.go, line 3744 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-2Sns-xs2sH0fyCN:-KY5-2Sns-xs2sH0fyCO:b-m46l8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3744)):*
> ```Go
> 			// Count good replicas: those which are on a live node and, if there is
> 			// a leader, which are not too far behind.
> 			hasLeader := isLeader || raftStatus != nil && raftStatus.SoftState.Lead != 0
> ```

this seems fragile. in the general case, a node can't definitely know there is no leader. in other words, this will double count if the highest node ID doesn't know of the leader but one exists.

---

*[pkg/storage/store.go, line 3746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-xaF5k0JlHc-ex7Z:-KY5-xaGERoTB5T0nHXo:b-ufkmqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3746)):*
> ```Go
> 			hasLeader := isLeader || raftStatus != nil && raftStatus.SoftState.Lead != 0
> 			var goodReplicas int
> 			highestIdx := -1
> ```

ouch, this code is getting really hard to parse.

aren't `desc.Replicas` guaranteed to be sorted? seems like there's an opportunity to simplify here based on that.

---

*[pkg/storage/store.go, line 3768 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY50m1R1c1iNuX3_fiY:-KY50m1R1c1iNuX3_fiZ:bc679r3) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3768)):*
> ```Go
> 			}
> 
> 			// If this replica is the leader or there isn't a leader and it is the
> ```

am i missing something? this doesn't check that there isn't a leader.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/storage/store.go, line 3744 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-2Sns-xs2sH0fyCN:-KY52ZfHReXQUSeWoX_W:b-1fcrxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3744)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this seems fragile. in the general case, a node can't definitely know there is no leader. in other words, this will double count if the highest node ID doesn't know of the leader but one exists.
</blockquote></details>

That's true, but I think that is the rare scenario as it requires the highest replica ID to be partitioned. The more common cases are that the replica knows the leader or the range is below quorum and there is no leader. I'll add a comment about the double counting in the partitioned case.

---

*[pkg/storage/store.go, line 3746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-xaF5k0JlHc-ex7Z:-KY52nzz7WlV2-oK0D22:b8exufw) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3746)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ouch, this code is getting really hard to parse.

aren't `desc.Replicas` guaranteed to be sorted? seems like there's an opportunity to simplify here based on that.
</blockquote></details>

I think `desc.Replicas` is sorted, but we don't rely on that guarantee elsewhere. Sure would be a shame to add that requirement here.

---

*[pkg/storage/store.go, line 3760 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY53TNkhl1MN9v5Pwf6:-KY53TNkhl1MN9v5Pwf7:b-7gh5hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3760)):*
> ```Go
> 						// can process thousands of ops/sec, so a replica that is 1000 Raft
> 						// log entries behind is fairly current.
> 						const behindThreshold = 1000
> ```

I've set this threshold down to 100 as 1000 appears to have been a bit high and seems to have been the source of the fall off in the 6m node outage under-replicated graph.

---

*[pkg/storage/store.go, line 3768 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY50m1R1c1iNuX3_fiY:-KY52wGbaBkUve4wOHrX:b-hrht51) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3768)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

am i missing something? this doesn't check that there isn't a leader.
</blockquote></details>

We only set `hightestIdx` if `!hasLeader`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/storage/store.go, line 3773 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5c4D0tqY_CkRVpUQj:-KY5c4D1usnQU1lSCdVq:bcr8xag) ([raw file](https://github.com/cockroachdb/cockroach/blob/879b4914f54b50b67a5e5156ba4262ec204a4327/pkg/storage/store.go#L3773)):*
> ```Go
> 						// hear otherwise. This is a bit of a hack.
> 						const behindThreshold = 100
> 						if progress.Match == 0 ||
> ```

Tracked down some flakiness in the metric to recent leadership changes. This line of code makes the metric stable, though it is a bit unfortunate as there are likely other reasons `progress.Match` can be 0. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by tamird** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-xaF5k0JlHc-ex7Z:-KY5hKugRAQmNvHpLjK1:bowvdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3746)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think `desc.Replicas` is sorted, but we don't rely on that guarantee elsewhere. Sure would be a shame to add that requirement here.
</blockquote></details>

I guess. Seems very very unlikely to change. By the way, it's not even necessary that it's sorted, only that the order is stable and consistent.

Seems worth it to me, given how hard this is to read now, but I won't insist.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/storage/store.go, line 3746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-xaF5k0JlHc-ex7Z:-KY6bYHF4gJ5yD-s3tMQ:bob7j1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3746)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I guess. Seems very very unlikely to change. By the way, it's not even necessary that it's sorted, only that the order is stable and consistent.

Seems worth it to me, given how hard this is to read now, but I won't insist.
</blockquote></details>

Cleanups come with risks. I've added a TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3773 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5c4D0tqY_CkRVpUQj:-KY7APBwDVJ_6w7EySC9:b-v83x2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/879b4914f54b50b67a5e5156ba4262ec204a4327/pkg/storage/store.go#L3773)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Tracked down some flakiness in the metric to recent leadership changes. This line of code makes the metric stable, though it is a bit unfortunate as there are likely other reasons `progress.Match` can be 0. PTAL.
</blockquote></details>

@bdarnell should probably have a look at this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by bdarnell** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-xaF5k0JlHc-ex7Z:-KY7A9EGAGTOeXNtXPv1:b-8nlk8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3746)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Cleanups come with risks. I've added a TODO.
</blockquote></details>

`desc.Replicas` is not currently sorted (new replicas are always added at the end, but we don't preserve order on removal). Since it's persisted, changing this is more trouble than it's worth. 

I think we could clean this up a bit by having one loop that just computes highestIdx instead of doing two different things in one loop.

---

*[pkg/storage/store.go, line 3760 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY53TNkhl1MN9v5Pwf6:-KY787-PBLMTQW9RtsUB:bcb64od) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3760)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I've set this threshold down to 100 as 1000 appears to have been a bit high and seems to have been the source of the fall off in the 6m node outage under-replicated graph.
</blockquote></details>

There are other fields we could use instead of heuristics on `Match`. `progress.State == raft.ProgressStateReplicate` is almost what we want. Even better would be `!progress.isPaused()` (which means "do we believe this node is up to date enough to send it new entries as they are proposed", and combines information about snapshots and MaxInFlightMsgs). I'll send a PR upstream to export this method. 

---

*[pkg/storage/store.go, line 3773 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5c4D0tqY_CkRVpUQj:-KY7COBn-z_27U3FvaL-:b-fcdbbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/879b4914f54b50b67a5e5156ba4262ec204a4327/pkg/storage/store.go#L3773)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

@bdarnell should probably have a look at this.
</blockquote></details>

I think it could use a comment that even though a Match of 0 could indicate a down node, the combination of Match 0 and livenessMap means it's more likely up. The logic is fine.

---

*[pkg/storage/store.go, line 3772 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY74Yzn7GZR1h2ZBJJh:-KY74Yzn1XGx8bsCA_o-:b-7h1jyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fdedf7dc2d0a2c90cc5cf51171f8b275955affc/pkg/storage/store.go#L3772)):*
> ```Go
> 			// such as determination of whether the current node should be recording
> 			// metrics (i.e. is the leader or highest live replica ID).
> 			hasLeader := isLeader || raftStatus != nil && raftStatus.SoftState.Lead != 0
> ```

This `isLeader ||` looks redundant. 

---

*[pkg/storage/store.go, line 3777 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY75pe-9QBCtgOqMcvN:-KY75pe-BIBKOPTxRo1c:b-l16pkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fdedf7dc2d0a2c90cc5cf51171f8b275955affc/pkg/storage/store.go#L3777)):*
> ```Go
> 			for i, rd := range desc.Replicas {
> 				if livenessMap[rd.NodeID] {
> 					if !hasLeader &&
> ```

We can compute the highestIdx whether we have a leader or not; probably better to avoid the condition here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-:-KY7CltKCfKDQuoAB9Kv:bnfp4nl)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-xaF5k0JlHc-ex7Z:-KY7D85I9zlgZhsv2eFd:bp4050x) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3746)):*
> desc.Replicas is not currently sorted (new replicas are always added at the end, but we don't preserve order on removal). Since it's persisted, changing this is more trouble than it's worth.

wouldn't relying on being first in the replicas slice still work?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-xaF5k0JlHc-ex7Z:-KY7GXzy7a6X6bKCVLVl:b4axmmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3746)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> desc.Replicas is not currently sorted (new replicas are always added at the end, but we don't preserve order on removal). Since it's persisted, changing this is more trouble than it's worth.

wouldn't relying on being first in the replicas slice still work?
</blockquote></details>

There's a bit of a race (around the time of a removal replicas will see different values in that position), but I guess it's no worse than the similar race around a newly-added replica at the highest index.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-xaF5k0JlHc-ex7Z:-KY9V_mM5p7_yEUeA7vc:bnn54qe) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3746)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

There's a bit of a race (around the time of a removal replicas will see different values in that position), but I guess it's no worse than the similar race around a newly-added replica at the highest index.
</blockquote></details>

Six of one, half a dozen of the other.

---

*[pkg/storage/store.go, line 3772 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY74Yzn7GZR1h2ZBJJh:-KY9UjrC2eT-Wg92ngjB:b-4uskvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fdedf7dc2d0a2c90cc5cf51171f8b275955affc/pkg/storage/store.go#L3772)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This `isLeader ||` looks redundant. 
</blockquote></details>

Yeah. I was being conservative. Are we guaranteed that the leader has `SoftState.Lead != 0`?

---

*[pkg/storage/store.go, line 3777 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY75pe-9QBCtgOqMcvN:-KY9VTSN2aPaXotMz9TL:b-beqciu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fdedf7dc2d0a2c90cc5cf51171f8b275955affc/pkg/storage/store.go#L3777)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We can compute the highestIdx whether we have a leader or not; probably better to avoid the condition here.
</blockquote></details>

True, but then I'd have to include `!hasLeader` in the condition below. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 3760 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY53TNkhl1MN9v5Pwf6:-KY9Vtts73ZZGqhp7o2c:b-73agr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3760)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

There are other fields we could use instead of heuristics on `Match`. `progress.State == raft.ProgressStateReplicate` is almost what we want. Even better would be `!progress.isPaused()` (which means "do we believe this node is up to date enough to send it new entries as they are proposed", and combines information about snapshots and MaxInFlightMsgs). I'll send a PR upstream to export this method. 
</blockquote></details>

I'm going to merge this and follow up with another PR to cleanup some of the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 3746 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY5-xaF5k0JlHc-ex7Z:-KYAr-Ri5lQ0uN1E3t8S:b-p4qci8) ([raw file](https://github.com/cockroachdb/cockroach/blob/67af137b2c25afc05dccf343a1ccc188ee473906/pkg/storage/store.go#L3746)):*
> Six of one, half a dozen of the other.

Right, except that the "the other" lets us write simpler code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*

**Comment by bdarnell** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 3772 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983#-KY74Yzn7GZR1h2ZBJJh:-KYC3MVfwMan0_1a7pwK:btpgocd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fdedf7dc2d0a2c90cc5cf51171f8b275955affc/pkg/storage/store.go#L3772)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yeah. I was being conservative. Are we guaranteed that the leader has `SoftState.Lead != 0`?
</blockquote></details>

Yes, the leader always [sets](https://github.com/coreos/etcd/blob/40f0193c4cee6f3e053143772056ef9b3124dba8/raft/raft.go#L606-L607) its own id in the `Lead` field.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/11983)*
